### PR TITLE
Refactor "app plugin loader" Gradle Plugin so it can be applied using the declarative `plugins {}` block

### LINKED ISSUE #125009: Load "flutter loader plugin" using Gradle's `plugins {}` block, rather than imperative `apply`
## Current situation

The default, `flutter create`d app has the `android/settings.gradle` file with the following content:

```gradle
include ":app"

def localPropertiesFile = new File(rootProject.projectDir, "local.properties")
def properties = new Properties()

assert localPropertiesFile.exists()
localPropertiesFile.withReader("UTF-8") { reader -> properties.load(reader) }

def flutterSdkPath = properties.getProperty("flutter.sdk")
assert flutterSdkPath != null, "flutter.sdk not set in local.properties"
apply from: "$flutterSdkPath/packages/flutter_tools/gradle/app_plugin_loader.gradle"
```

This line: 

```gradle
apply from: "$flutterSdkPath/packages/flutter_tools/gradle/app_plugin_loader.gradle"
```

imperatively applies the [`app_plugin_loader.gradle` script](https://github.com/flutter/flutter/blob/3.7.11/packages/flutter_tools/gradle/app_plugin_loader.gradle) to load the Flutter plugins (that have Android implementation).

## Preferred situation

We could create a new Gradle plugin (strictly speaking: a new Gradle settings plugin) to make the logic more declarative. I'm thinking about:

> Note: this is a new default `settings.gradle` (with changes made in #123511):

```groovy
pluginManagement {
    def flutterSdkPath = {
        def properties = new Properties()
        file("local.properties").withInputStream { properties.load(it) }
        def flutterSdkPath = properties.getProperty("flutter.sdk")
        assert flutterSdkPath != null, "flutter.sdk not set in local.properties"
        return flutterSdkPath
    }
    settings.ext.flutterSdkPath = flutterSdkPath()

    // Flutter Gradle Plugin ships together with the Flutter SDK
    includeBuild("${settings.ext.flutterSdkPath}/packages/flutter_tools/gradle")

    plugins {
        // Add the default Flutter Gradle Plugin
        id "dev.flutter.flutter-gradle-plugin" version "1.0.0" apply false

        // Add the Flutter plugin loader Gradle Plugin
        id "dev.flutter.flutter-plugin-loader" version "1.0.0" apply false
    }
}

include ':app'
```

### Docs

From Gradle docs ([Developing Custom Gradle Plugins](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:writing_a_simple_plugin)):

```kotlin
class GreetingPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        project.task("hello") {
            doLast {
                println("Hello from the GreetingPlugin")
            }
        }
    }
}

// Apply the plugin
apply<GreetingPlugin>()
```

> Note that the [Plugin](https://docs.gradle.org/current/javadoc/org/gradle/api/Plugin.html) class is a generic type. This example has it receiving the [Project](https://docs.gradle.org/current/dsl/org.gradle.api.Project.html) type as a type parameter. A plugin can instead receive a parameter of type [Settings](https://docs.gradle.org/current/dsl/org.gradle.api.initialization.Settings.html), in which case the plugin can be applied in a settings script, or a parameter of type [Gradle](https://docs.gradle.org/current/dsl/org.gradle.api.invocation.Gradle.html), in which case the plugin can be applied in an initialization script.

## Use case

This doesn't have any immediate benefit to developers. It'll make us more compliant with best practices, though.

The biggest benefit that I see is that it'll allow us to migrate the `app_plugin_loader` from Groovy to Kotlin without users having to change the path in `settings.gradle` manually (somewhat related issue: #121541).

Other Gradle-related improvement issues:
- #121552
- #123934
- #124167

### PR DESCRIPTION
**Pull request by bartekpacia**
This PR fixes #125009.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (bartekpacia) [inline packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:1]: This plugin is much smaller than the [main Flutter Gradle Plugin](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy).

I think it's a good candidate for Kotlin migration - see #121541

**Comment by bartekpacia** [comment]
I gotta fix the tests first, ugh.

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:1]: cc @reidbaker @gnprice

File: packages/flutter_tools/gradle/build.gradle.kts
Comment (reidbaker) [inline packages/flutter_tools/gradle/build.gradle.kts:19]: If it does not cause issues consider making this 7.4.2 which I think is the highest 7.* version published.

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:1]: Not opposed, Mild preference towards doing conversions in prs independent of something like changing the structure.

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (reidbaker) [inline packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:28]: Nit: use named constant here and link to documentation for where maintainers can read more about it.

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (reidbaker) [inline packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:26]: Great comment.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (reidbaker) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:1059]: Why stop using a constant and move to a string? that feels like a regression.

File: packages/flutter_tools/gradle/src/main/kotlin/flutter.gradle.kts
Comment (reidbaker) [inline packages/flutter_tools/gradle/src/main/kotlin/flutter.gradle.kts:7]: Thank you for continually making this file smaller by extracting functionality into their own files.

**Review by reidbaker — COMMENTED** [review]
Marking this "request changes" but in a comment. I am about to leave for 2 weeks of vacation  and don't want to hold this up if we can get sufficient approval. 

Please dont merge without @stuartmorgan or @dnfield giving an LGTM.  They are other experienced flutter team members with some comfort in android.

**Review by dnfield — DISMISSED** [review]

File: packages/flutter_tools/gradle/build.gradle.kts
Comment (johnmccutchan) [reply packages/flutter_tools/gradle/build.gradle.kts:19]: +1

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (johnmccutchan) [inline packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:12]: Might be nice if you blocked off the section of the code that needs to be kept in sync with module_plugin_loader.gradle:

// If any of the following code changes, you must update path/to/file
... code ...
// End of code that must be kept in sync with path/to/file

And you probably want a similar comment to be added in path/to/file pointing back at app_plugin_loader.groovy.

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (johnmccutchan) [inline packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:26]: Please add a test that confirms that such plugins are skipped.

**Comment by bartekpacia** [comment]
Wow thank you all for quick reviews! But tbh this PR should've been a draft, tests are failing are I made some very ugly things, as you've already seen :)

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:1]: Sure, I'll convert it into Kotlin in another, future PR.

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:26]: Not mine :) I mostly copy-pasted this haha

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:1059]: I did this because for some unknown to me reason, the `com.android.build.OutputFile` import works only at compile time, but throws a ClassNotFound exception at runtime.

`OutputFile.ABI` is the only usage of this import, so I made a little hack and replaced it with the string "ABI" just to get it working.

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:28]: This is copy-pasted, but I'm more than happy to make improvements like this (after I get all the tests to pass...)

File: packages/flutter_tools/gradle/build.gradle.kts
Comment (bartekpacia) [reply packages/flutter_tools/gradle/build.gradle.kts:55]: Yup I agree with all of that. This is a temporary hack because for some reason, I started getting duplicate class error on `FlutterPlugin`.

Will debug and once I find a solution, I'll get rid of this.

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:26]: Good idea - will definitely do it after I get tests to pass.

**Comment by bartekpacia** [comment]
Some debugging notes re: Why `java.lang.NoClassDefFoundError: com/android/build/OutputFile` is thrown?

I found this issue and [this comment](https://github.com/JLLeitschuh/ktlint-gradle/issues/47#issuecomment-353948863). This looked promising, so I decided to convert the Android part of my Flutter app to use the declarative-style plugin apply everywhere (just like native Android projects created with Android studio look like).
Unfortunately, that didn't fix the problem. I'm still getting the error:

```bash
$ ./gradlew :app:assembleDebug
> Task :app:assembleDebug FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:assembleDebug'.
> com/android/build/OutputFile
```

What's also strange is that this problem started occurring in this PR, which _does not touches the main Flutter Gradle Plugin_. Or does it? TO VERIFY: Does this error occur on current `master`?

**Comment by reidbaker** [comment]
@bartekpacia I am back from a set of vacations and traveling for google io. Let me know if/when I can give a helpful review by pinging me. Sorry for the lack of response up until now.

**Comment by bartekpacia** [comment]
Hi @reidbaker, thanks for willingness to help. Unfortunately last few weeks have been quite busy (work, studies, FlutterCon), but now I hope to find some more time to work on this :)

**Comment by gnprice** [comment]
I did a quick search in the Gradle issue tracker, and it looks like the exception you're seeing may be this Gradle bug:
https://github.com/gradle/gradle/issues/23709

Unfortunately it's not clear how to proceed from there. That bug sure looks like a bug, and if that's indeed the cause here then it's not because of anything your code is doing wrong. There's no evident workaround either.

I guess that issue does offer one clear option for making progress: go start learning the Gradle codebase :smile: by debugging that Gradle bug, and report your findings in the thread and/or send a PR to fix it.

**Comment by TWiStErRob** [comment]
```patch
diff --git a/examples/hello_world/android/build.gradle b/examples/hello_world/android/build.gradle
index 2de2062357..ce30c1b6eb 100644
--- a/examples/hello_world/android/build.gradle
+++ b/examples/hello_world/android/build.gradle
@@ -14,7 +14,6 @@ buildscript {
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:7.2.0'
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
     }
 
diff --git a/examples/hello_world/android/settings.gradle b/examples/hello_world/android/settings.gradle
index 4eb2a21d71..8f4d6f73e0 100644
--- a/examples/hello_world/android/settings.gradle
+++ b/examples/hello_world/android/settings.gradle
@@ -19,6 +19,11 @@ pluginManagement {
     // Flutter Gradle Plugin ships together with the Flutter SDK
     includeBuild("${settings.ext.flutterSdkPath}/packages/flutter_tools/gradle")
 
+    repositories {
+        gradlePluginPortal()
+        google()
+    }
+
     plugins {
         // Flutter Gradle Plugin's ID is defined in /packages/flutter_tools/gradle/build.gradle.kts.
         // We set `apply false`, because we don't want to apply the plugin to
@@ -27,8 +32,16 @@ pluginManagement {
     }
 }
 
+buildscript {
+    configurations.classpath {
+        resolutionStrategy.activateDependencyLocking()
+    }
+    //dependencyLocking.lockFile = file("${rootProject.projectDir}/settings.lockfile") // default is settings-gradle.lockfile
+}
+
 plugins {
     id "dev.flutter.flutter-plugin-loader" version "1.0.0"
+    id "com.android.application" version "7.2.0" apply false
 }
 
 include ':app'
```

**Comment by bartekpacia** [comment]
These CI failures all seem to be caused by `jcenter.bintray.com` not being available. Example:

<details><summary>Details</summary>

```
Expected: a process with exit code 0
    Actual: <Instance of 'ProcessResult'>
     Which: Actual exitCode was 1
            Actual stderr:
            
            FAILURE: Build failed with an exception.
            
            * What went wrong:
            Could not resolve all files for configuration 'classpath'.
            > Could not download txw2-2.3.2.jar (org.glassfish.jaxb:txw2:2.3.2)
               > Could not get resource 'https://plugins.gradle.org/m2/org/glassfish/jaxb/txw2/2.3.2/txw2-2.3.2.jar'.
                  > Could not GET 'https://jcenter.bintray.com/org/glassfish/jaxb/txw2/2.3.2/txw2-2.3.2.jar'.
                     > peer not authenticated
            > Could not download checker-qual-3.5.0.jar (org.checkerframework:checker-qual:3.5.0)
               > Could not get resource 'https://plugins.gradle.org/m2/org/checkerframework/checker-qual/3.5.0/checker-qual-3.5.0.jar'.
                  > Could not GET 'https://jcenter.bintray.com/org/checkerframework/checker-qual/3.5.0/checker-qual-3.5.0.jar'.
                     > peer not authenticated
            > Could not download netty-codec-socks-4.1.52.Final.jar (io.netty:netty-codec-socks:4.1.52.Final)
               > Could not get resource 'https://plugins.gradle.org/m2/io/netty/netty-codec-socks/4.1.52.Final/netty-codec-socks-4.1.52.Final.jar'.
                  > Could not GET 'https://jcenter.bintray.com/io/netty/netty-codec-socks/4.1.52.Final/netty-codec-socks-4.1.52.Final.jar'.
                     > peer not authenticated
            
            * Try:
            > Run with --stacktrace option to get the stack trace.
            > Run with --info or --debug option to get more log output.
            > Run with --scan to get full insights.
            
            * Get more help at https://help.gradle.org/
            
            BUILD FAILED in 9s
```

</details>

**Comment by bartekpacia** [comment]
These CI failures are very cryptic and random, I've tried to reproduce them locally without success. I'd be grateful if someone looked at them and shared ideas what could be causing them.

**Comment by reidbaker** [comment]
> > These CI failures are very cryptic and random, I've tried to reproduce them locally without success. I'd be grateful if someone looked at them and shared ideas what could be causing them.
> 
> Digging into a specific failure https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20tool_integration_tests_2_4/26918/overview I see a failed test that was written by me to have gradle expose what version of java it was running with [2723266](https://github.com/flutter/flutter/commit/27232664668033751aad6faa0f7b37ef1bbac0ed).
> 
> ```
> 02:57 +3 ~3 -1: test/integration.shard/android_gradle_java_version_test.dart: gradle task exists named javaVersion that prints jdk version [E]                                      
> ```
> 
> In sudo code the test is flutter create, config android project, run `gradlew -q javaVersion`

Test file is here if it helps 
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/android_gradle_java_version_test.dart

**Comment by bartekpacia** [comment]
Thanks for taking a look @reidbaker :) Actually I found that test, but still, I've literally no idea what could be causing it to fail:

```
* What went wrong:
            Could not resolve all files for configuration 'classpath'.
            > Could not download netty-transport-4.1.52.Final.jar (io.netty:netty-transport:4.1.52.Final)
               > Could not get resource 'https://plugins.gradle.org/m2/io/netty/netty-transport/4.1.52.Final/netty-transport-4.1.52.Final.jar'.
                  > Could not GET 'https://jcenter.bintray.com/io/netty/netty-transport/4.1.52.Final/netty-transport-4.1.52.Final.jar'.
                     > peer not authenticated
```

> Thanks Gradle, but pretty much all other tests pass, so...?

**Comment by reidbaker** [comment]
> Thanks for taking a look @reidbaker :) Actually I found that test (`android_gradle_java_version_test.dart`), but still, I've literally no idea what could be causing it to fail:
> 
> ```
> * What went wrong:
>             Could not resolve all files for configuration 'classpath'.
>             > Could not download netty-transport-4.1.52.Final.jar (io.netty:netty-transport:4.1.52.Final)
>                > Could not get resource 'https://plugins.gradle.org/m2/io/netty/netty-transport/4.1.52.Final/netty-transport-4.1.52.Final.jar'.
>                   > Could not GET 'https://jcenter.bintray.com/io/netty/netty-transport/4.1.52.Final/netty-transport-4.1.52.Final.jar'.
>                      > peer not authenticated
> ```
> 
> Thanks Gradle, but pretty much all other tests pass, why only this one encounters this error?

FWIW that asset downloaded on my work machine so the url seems valid.

**Comment by bartekpacia** [comment]
Works locally on my machine as well - how weird.

File: packages/flutter_tools/templates/app_shared/android.tmpl/settings.gradle.tmpl
Comment (TWiStErRob) [inline packages/flutter_tools/templates/app_shared/android.tmpl/settings.gradle.tmpl:16]: Re https://github.com/flutter/flutter/pull/127897#issuecomment-1715906594

Long shot: is it possible the tests are running in parallel in isolated environments, but from the same-ish IP range, so the shards are DDoS'ing gradlePluginPortal/JCenter and get rate limited with a strange error message? Or simply JCenter is flaky enough to fail with a high enough percentage that there's always a failure?

Either way, @you can probably prevent failures if you [put](https://docs.gradle.org/current/userguide/plugins.html#sec:custom_plugin_repositories) `mavenCentral()` in front of `gradlePluginPortal` to ensure that less artifacts are hitting JCenter. The artifact in question is available in MavenCentral too: https://mvnrepository.com/artifact/io.netty/netty-transport/4.1.52.Final

The weird thing is: why this PR? The default is `repositories { gradlePluginPortal() }` so this didn't change much... although, the 2 new plugins below might result in more files being resolved?

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: This is fixed, i.e. I went back to using a constant.

File: packages/flutter_tools/gradle/src/main/kotlin/FlutterPluginKts.kt
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/kotlin/FlutterPluginKts.kt]: This code is copy-pasted from the previous file. I'm happy to do these changes in a future PR.

File: packages/flutter_tools/templates/app_shared/android.tmpl/settings.gradle.tmpl
Comment (bartekpacia) [reply packages/flutter_tools/templates/app_shared/android.tmpl/settings.gradle.tmpl]: Sure, after CI passes for [`4109747`](https://github.com/flutter/flutter/pull/127897/commits/410974794f249d2d73fec676a3ebd47a2e69304d) (what a funny commit number btw!) I'm happy to do it.

File: packages/flutter_tools/gradle/build.gradle.kts
Comment (bartekpacia) [reply packages/flutter_tools/gradle/build.gradle.kts]: I tried in [`0e7e128`](https://github.com/flutter/flutter/pull/127897/commits/0e7e1282bf5b7032dddd0cf1c7718264f535fb99), but CI failed because of those changes.

Let's bump this version in some another PR.

File: packages/flutter_tools/gradle/build.gradle.kts
Comment (TWiStErRob) [inline packages/flutter_tools/gradle/build.gradle.kts]: If you're targeting Java 8, you might get into conflicting versions between the project and AGP artifacts. See [here](https://github.com/TWiStErRob/net.twisterrob.gradle/blob/0.16/gradle/plugins/src/main/kotlin/net/twisterrob/gradle/build/compilation/JavaCompatibilityPlugin.kt) for the latest variant of this problem, or [here](https://github.com/TWiStErRob/net.twisterrob.gradle/pull/364/files#diff-e2e837f5947404283a07c9a53beef24153ed15da97493c0dba0ac80238f7d7a6) for an older version that matches your tried upgrade.

Comment (bartekpacia) [reply packages/flutter_tools/gradle/build.gradle.kts]: thanks, it'll for sure come in useful in a future PR when Gradle & AGP versions will be bumped.

btw the first links 404's

File: packages/flutter_tools/templates/app_shared/android.tmpl/settings.gradle.tmpl
Comment (bartekpacia) [reply packages/flutter_tools/templates/app_shared/android.tmpl/settings.gradle.tmpl:16]: Reverted [`29b01c3`](https://github.com/flutter/flutter/commit/29b01c39ad9124dea6d4c493352fe379a9ba8c8e) in [`a39fc08`](https://github.com/flutter/flutter/pull/127897/commits/a39fc08fd92d646d1b8f44b8a4933dd872524990) and the same error appeard.

File: packages/flutter_tools/gradle/build.gradle.kts
Comment (TWiStErRob) [reply packages/flutter_tools/gradle/build.gradle.kts]: Yeah, defo separate PR, it's a tricky one; fixed link, sorry.

File: packages/flutter_tools/templates/app_shared/android.tmpl/settings.gradle.tmpl
Comment (TWiStErRob) [reply packages/flutter_tools/templates/app_shared/android.tmpl/settings.gradle.tmpl]: Thanks, so that confirms to me that gPP/jcenter can't handle missing artifacts (at least not a lot of them at once).
In the CI of https://github.com/flutter/flutter/pull/127897/checks?check_run_id=16762079212
this was the error https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8770052323773862369/+/u/run_test.dart_for_tool_integration_tests_shard_and_subshard_1_4/test_stdout#L4443_17
which is a different artifact than the original Netty one.
I might try to reproduce it in some way and report to Gradle if possible.

(re funny commit number: https://github.com/not-an-aardvark/lucky-commit)

Comment (TWiStErRob) [reply packages/flutter_tools/templates/app_shared/android.tmpl/settings.gradle.tmpl]: I tried sending 1000 requests with 200-1000 parallelism for a `.pom`, but didn't get this "peer" error. Only socket closed and connection reset.

**Comment by bartekpacia** [comment]
The failing `Mac web_tool_tests` is unrelated to this PR, seems to be flaky.

**Comment by bartekpacia** [comment]
This time it's `Windows build_tests_4_5` simply timing out.

**Comment by bartekpacia** [comment]
Finally, green :) @reidbaker @dnfield politely pinging for a review

File: packages/flutter_tools/gradle/settings.gradle.kts
Comment (reidbaker) [inline packages/flutter_tools/gradle/settings.gradle.kts:2]: I think this only applies to our own plugin as defined in flutter tools and does not apply to consumers of flutter is that correct?

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:26]: can you resolve this when you add this test?

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:28]: nit: now that the tests pass lets use a constant here.

**Review by reidbaker — APPROVED** [review]
Approve mod nits

File: packages/flutter_tools/gradle/settings.gradle.kts
Comment (bartekpacia) [reply packages/flutter_tools/gradle/settings.gradle.kts:2]: yup, correct

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:26]: I'm not sure where I should get started with writing such test.

I searched and found devicelab test called [`plugin_test`](https://github.com/flutter/flutter/blob/master/dev/devicelab/bin/tasks/plugin_test.dart), and run it locally - it passed.


```
../../bin/cache/dart-sdk/bin/dart bin/test_runner.dart test -t plugin_test
```

I commented out the snippet that @johnmccutchan suggests to write tests for. I did:

```groovy
//def needsBuild = androidPlugin.containsKey(nativeBuildKey) ? //androidPlugin[nativeBuildKey] : true
//if (!needsBuild) {
//    return
//}
```

and then run `plugin_test` again, and it failed consistently across 3 executions.

<details><summary>crash logs</summary>

```
[2023-09-14 22:23:59.607828] [STDOUT] stderr: [   +1 ms] FAILURE: Build failed with an exception.
[2023-09-14 22:23:59.607867] [STDOUT] stderr: [   +1 ms] * What went wrong:
[2023-09-14 22:23:59.607914] [STDOUT] stderr: [        ] assert pluginDirectory.exists()
[2023-09-14 22:23:59.607925] [STDOUT] stderr: [        ]        |               |
[2023-09-14 22:23:59.607934] [STDOUT] stderr: [        ]        |               false
[2023-09-14 22:23:59.607943] [STDOUT] stderr: [        ]        /private/var/folders/z8/vl6s_16515s6h_j8z7bs1tmm0000gn/T/flutter_devicelab_plugin_test.3X0VcH/plugintest/android
[2023-09-14 22:23:59.607953] [STDOUT] stderr: [        ] * Try:
[2023-09-14 22:23:59.607989] [STDOUT] stderr: [        ] > Run with --debug option to get more log output.
[2023-09-14 22:23:59.608168] [STDOUT] stderr: [        ] > Run with --scan to get full insights.
[2023-09-14 22:23:59.608185] [STDOUT] stderr: [        ] * Exception is:
[2023-09-14 22:23:59.608195] [STDOUT] stderr: [        ] Assertion failed:
[2023-09-14 22:23:59.608219] [STDOUT] stderr: [        ] assert pluginDirectory.exists()
[2023-09-14 22:23:59.608233] [STDOUT] stderr: [        ]        |               |
[2023-09-14 22:23:59.608242] [STDOUT] stderr: [        ]        |               false
[2023-09-14 22:23:59.608398] [STDOUT] stderr: [        ]        /private/var/folders/z8/vl6s_16515s6h_j8z7bs1tmm0000gn/T/flutter_devicelab_plugin_test.3X0VcH/plugintest/android
```

</details> 


So I think it's not necessary to add such test in this PR, esp. that this code is not mine – I just copied it into a new location.

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy]: > link to documentation for where maintainers can read more about it.

That's a good idea but I searched for a while but couldn't find any docs referring `native_build`. It's defined here:

https://github.com/flutter/flutter/blob/ff10c52ad6de098b4946f9ef33fdde8ebd5bc594/packages/flutter_tools/lib/src/flutter_plugins.dart#L106

File: packages/flutter_tools/gradle/settings.gradle.kts
Comment (reidbaker) [reply packages/flutter_tools/gradle/settings.gradle.kts:2]: Perfect!

File: packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy
Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy:30]: The test failing with the code commented out is good enough for me.

**Review by gmackall — APPROVED** [review]
This LGTM. 

I'll hold off on autosubmitting until tomorrow, since it's a relatively big change to our templates and I want to double check with @reidbaker if there was anything he wanted to confirm before merging (he's ooo today).

**Comment by bartekpacia** [comment]
Hey @gmackall, thanks for reviewing - any chance of landing it today?

**Comment by gmackall** [comment]
Dismissing review from @dnfield as stale (all points were addressed), so I can get autosubmit to stick.

**Pull request merged**
